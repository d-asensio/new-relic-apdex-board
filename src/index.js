
import './styles/index.scss'

import { DashboardController } from './controllers'
import { DashboardService } from './services'
import { DashboardView } from './views'

const controller = new DashboardController(
  new DashboardService(),
  new DashboardView()
)

function renderApp () {
  const appRoot = document.getElementById('app')
  controller.renderToDom(appRoot)
}

;(async function IIFE () {
  await controller.init()

  if (document.readyState === 'complete') {
    renderApp()
  } else {
    document.addEventListener(
      'DOMContentLoaded',
      () => renderApp()
    )
  }
})()
