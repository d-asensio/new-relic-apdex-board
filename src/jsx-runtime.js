export default function createElement (tag, props, ...children) {
  const element = document.createElement(tag)

  if (props !== null) {
    _addPropsToElement(props, element)
  }

  _appendChildrenToElement(children, element)

  return element
}

function _addPropsToElement (props, element) {
  const { className, ...attributes } = props

  _addAttributesToElement(attributes, element)

  if (className) {
    _addClassNameToElement(className, element)
  }
}

function _addAttributesToElement (attributes, element) {
  const attributeEntries = Object.entries(attributes)

  for (const [attrName, attrValue] of attributeEntries) {
    _attributeSupportedByElementOrThrow(attrName, element)

    element.setAttribute(attrName, attrValue)
  }
}

function _attributeSupportedByElementOrThrow (attributeName, element) {
  if (!_isAttributeSupportedByElement(attributeName, element)) {
    throw new Error(
      `A "${element.tagName.toLowerCase()}" element do not support the attribute "${attributeName}".`
    )
  }
}

function _isAttributeSupportedByElement (attributeName, element) {
  if (_isDataAttribute(attributeName)) return true
  if (_isAriaAttribute(attributeName)) return true

  return attributeName in element
}

function _isDataAttribute (attributeName) {
  return attributeName.startsWith('data-')
}

function _isAriaAttribute (attributeName) {
  return attributeName.startsWith('aria-')
}

function _addClassNameToElement (className, element) {
  const sanitizedClassNames = _sanitizeClassName(className)

  for (const sanitizedClass of sanitizedClassNames) {
    element.classList.add(sanitizedClass)
  }
}

function _sanitizeClassName (className) {
  return className
    .split(' ')
    .filter(Boolean)
}

function _appendChildrenToElement (children, element) {
  for (const child of children) {
    if (Array.isArray(child)) {
      _appendChildrenToElement(child, element)
    } else {
      _appendChildToElement(child, element)
    }
  }
}

function _appendChildToElement (child, element) {
  const safeChildNode = _createSafeChildNode(child, element)

  element.appendChild(safeChildNode)
}

function _createSafeChildNode (child) {
  if (child instanceof HTMLElement || child instanceof Text) {
    return child
  }

  return document.createTextNode(child)
}
