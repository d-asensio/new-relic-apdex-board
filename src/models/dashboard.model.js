export class Dashboard {
  constructor ({ user }) {
    this.user = user

    this._hosts = new Map()
    this._apps = new Map()
  }

  addApp (app) {
    this._apps.set(app.id, app)
  }

  hasApp (appId) {
    return this._apps.has(appId)
  }

  getApp (appId) {
    return this._apps.get(appId)
  }

  addHost (host) {
    this._hosts.set(host.id, host)
  }

  hasHost (hostId) {
    return this._hosts.has(hostId)
  }

  getHost (hostId) {
    return this._hosts.get(hostId)
  }

  getAllHosts () {
    return Array.from(this._hosts.values())
  }
}
