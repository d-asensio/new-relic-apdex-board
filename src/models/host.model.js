export class Host {
  constructor ({ id }) {
    this.id = id
    this._apps = []
  }

  addApp (app) {
    this._insertAppInPosition(app)
  }

  removeApp (app) {
    this._removeAppFromPosition(app)
  }

  getTopApps (nTopApps) {
    return this._apps.slice(0, nTopApps)
  }

  _insertAppInPosition (app) {
    const appInsertionIndex = this._binarySearch(app)
    this._insertAppAtIndex(app, appInsertionIndex)
  }

  _removeAppFromPosition (app) {
    const appIndex = this._binarySearch(app)

    this._indexExistOrThrow(appIndex)
    this._removeAppAtIndex(appIndex)
  }

  _binarySearch (app) {
    let startIndex = 0
    let endIndex = this._apps.length - 1
    let currentIndex

    while (startIndex <= endIndex) {
      currentIndex = (startIndex + endIndex) / 2 | 0
      const iteratingApp = this._apps[currentIndex]

      if (app.compareTo(iteratingApp) < 0) {
        startIndex = currentIndex + 1
      } else if (app.compareTo(iteratingApp) > 0) {
        endIndex = currentIndex - 1
      } else {
        return currentIndex
      }
    }

    return Math.abs(~endIndex)
  }

  _insertAppAtIndex (app, insertionIndex) {
    this._apps.splice(insertionIndex, 0, app)
  }

  _removeAppAtIndex (deletionIndex) {
    this._apps.splice(deletionIndex, 1)
  }

  _indexExistOrThrow (index) {
    if (!this._indexExists(index)) {
      throw new Error(
        'The specified app do not exist in the host.'
      )
    }
  }

  _indexExists (index) {
    return index >= 0 && index < this._apps.length
  }
}
