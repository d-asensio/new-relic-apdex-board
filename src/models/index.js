export * from './application.model.js'
export * from './dashboard.model.js'
export * from './host.model.js'
