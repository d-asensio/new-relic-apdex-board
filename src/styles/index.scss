@import "reset";

@import "tokens/all";
@import "utilities/all";

@import "views/all";

body {
  background: #f5f5f5;
  
  @include sp-inset($space--large);
}
