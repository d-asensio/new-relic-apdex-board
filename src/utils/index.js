export * from './event-bus'
export * from './unique-id'
