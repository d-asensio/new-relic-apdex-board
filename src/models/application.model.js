import { uniqueId } from '../utils'

export class Application {
  constructor ({ name, apdex, version, contributors }) {
    this._id = uniqueId('app_')

    this._name = name
    this._apdex = apdex
    this._version = version
    this._contributors = contributors
  }

  get id () { return this._id }
  get name () { return this._name }
  get apdex () { return this._apdex }
  get version () { return this._version }
  get contributors () { return this._contributors }

  compareTo (comparee) {
    return this.apdex - comparee.apdex
  }

  static byApdexInDescOrder (appA, appB) {
    return appB.apdex - appA.apdex
  }
}
