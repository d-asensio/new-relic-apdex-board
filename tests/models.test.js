import { Application, Dashboard, Host } from '../src/models'

describe('Application model', () => {
  it('is constructed', () => {
    const app = new Application({
      name: 'A',
      version: 2,
      apdex: 55,
      contributors: [
        'John Doe'
      ]
    })

    expect(app).toBeInstanceOf(Application)
  })

  it('generates an id', () => {
    const app = new Application({
      name: 'A',
      version: 2,
      apdex: 55,
      contributors: [
        'John Doe'
      ]
    })

    expect(app.id).toBeDefined()
  })

  it('the id is not modificable', () => {
    const app = new Application({
      name: 'A',
      version: 2,
      apdex: 55,
      contributors: [
        'John Doe'
      ]
    })

    expect(() => {
      app.id = 'new_id'
    }).toThrowError()
  })

  it('has name', () => {
    const name = 'A'

    const app = new Application({
      name,
      version: 2,
      apdex: 55,
      contributors: [
        'John Doe'
      ]
    })

    expect(app.name).toBe(name)
  })

  it('the name is not modificable', () => {
    const app = new Application({
      name: 'A',
      version: 2,
      apdex: 55,
      contributors: [
        'John Doe'
      ]
    })

    expect(() => {
      app.name = 'new_name'
    }).toThrowError()
  })

  it('has version', () => {
    const version = 2

    const app = new Application({
      name: 'A',
      version,
      apdex: 55,
      contributors: [
        'John Doe'
      ]
    })

    expect(app.version).toBe(version)
  })

  it('the version is not modificable', () => {
    const app = new Application({
      name: 'A',
      version: 2,
      apdex: 55,
      contributors: [
        'John Doe'
      ]
    })

    expect(() => {
      app.version = 5
    }).toThrowError()
  })

  it('has apdex', () => {
    const apdex = 55

    const app = new Application({
      name: 'A',
      version: 2,
      apdex,
      contributors: [
        'John Doe'
      ]
    })

    expect(app.apdex).toBe(apdex)
  })

  it('the apdex is not modificable', () => {
    const app = new Application({
      name: 'A',
      version: 2,
      apdex: 55,
      contributors: [
        'John Doe'
      ]
    })

    expect(() => {
      app.apdex = 58
    }).toThrowError()
  })

  it('has contributors', () => {
    const contributors = [
      'John Doe'
    ]

    const app = new Application({
      name: 'A',
      version: 2,
      apdex: 55,
      contributors
    })

    expect(app.contributors).toBe(contributors)
  })

  it('contributors are not modificable', () => {
    const app = new Application({
      name: 'A',
      version: 2,
      apdex: 55,
      contributors: [
        'John Doe'
      ]
    })

    expect(() => {
      app.contributors = []
    }).toThrowError()
  })

  it('can be compared to another app', () => {
    const appA = new Application({
      name: 'A',
      version: 2,
      apdex: 55,
      contributors: []
    })

    const appB = new Application({
      name: 'B',
      version: 2,
      apdex: 60,
      contributors: []
    })

    const comparison = appA.compareTo(appB)

    expect(comparison).toBe(-5)
  })

  it('has a comparator that can be used to sort in descencent order', () => {
    const appA = new Application({
      name: 'A',
      version: 2,
      apdex: 55,
      contributors: []
    })

    const appB = new Application({
      name: 'B',
      version: 2,
      apdex: 60,
      contributors: []
    })

    const comparison = Application.byApdexInDescOrder(appA, appB)

    expect(comparison).toBe(5)
  })
})

describe('Dashboard model', () => {
  it('is constructed', () => {
    const dashboard = new Dashboard({
      user: 'some.one@newrelic.com'
    })

    expect(dashboard).toBeInstanceOf(Dashboard)
  })

  it('has user', () => {
    const user = 'some.one@newrelic.com'

    const dashboard = new Dashboard({
      user
    })

    expect(dashboard.user).toBe(user)
  })

  it('the user is not modificable', () => {
    const dashboard = new Dashboard({
      user: 'some.one@newrelic.com'
    })

    expect(() => {
      dashboard.user = 'some.other@newrelic.com'
    }).toThrowError()
  })

  it('adds an app', () => {
    const dashboard = new Dashboard({
      user: 'some.one@newrelic.com'
    })

    const app = new Application({
      name: 'A',
      version: 2,
      apdex: 55,
      contributors: []
    })

    dashboard.addApp(app)

    const hasApp = dashboard.hasAppId(app.id)

    expect(hasApp).toBe(true)
  })

  it('gets an app', () => {
    const dashboard = new Dashboard({
      user: 'some.one@newrelic.com'
    })

    const app = new Application({
      name: 'A',
      version: 2,
      apdex: 55,
      contributors: []
    })

    dashboard.addApp(app)

    const gotApp = dashboard.getAppById(app.id)

    expect(app).toBe(gotApp)
  })

  it('do not modify apps when adding', () => {
    const dashboard = new Dashboard({
      user: 'some.one@newrelic.com'
    })

    const app = new Application({
      name: 'A',
      version: 2,
      apdex: 55,
      contributors: []
    })

    Object.freeze(app)

    dashboard.addApp(app)
  })

  it('adds a host', () => {
    const dashboard = new Dashboard({
      user: 'some.one@newrelic.com'
    })

    const host = new Host({
      id: 'a2f3d.host.com'
    })

    dashboard.addHost(host)

    const hasHost = dashboard.hasHostId(host.id)

    expect(hasHost).toBe(true)
  })

  it('gets a host', () => {
    const dashboard = new Dashboard({
      user: 'some.one@newrelic.com'
    })

    const host = new Host({
      id: 'a2f3d.host.com'
    })

    dashboard.addHost(host)

    const gotHost = dashboard.getHostById(host.id)

    expect(gotHost).toBe(host)
  })

  it('gets multiple hosts', () => {
    const dashboard = new Dashboard({
      user: 'some.one@newrelic.com'
    })

    const hostA = new Host({
      id: 'a2f3d.hostA.com'
    })

    const hostB = new Host({
      id: 'a2f3d.hostB.com'
    })

    dashboard.addHost(hostA)
    dashboard.addHost(hostB)

    const allHosts = dashboard.getAllHosts()

    expect(allHosts).toStrictEqual([hostA, hostB])
  })

  it('do not modify hosts when adding', () => {
    const dashboard = new Dashboard({
      user: 'some.one@newrelic.com'
    })

    const host = new Host({
      id: 'a2f3d.host.com'
    })

    Object.freeze(host)

    dashboard.addHost(host)
  })
})

describe('Host model', () => {
  it('is constructed', () => {
    const host = new Host({
      id: 'a2f3d.host.com'
    })

    expect(host).toBeInstanceOf(Host)
  })

  it('return an empty top apps array when there are no apps', () => {
    const host = new Host({
      id: 'a2f3d.host.com'
    })

    expect(host.getTopApps(10)).toStrictEqual([])
  })

  it('do not modify an added app', () => {
    const host = new Host({
      id: 'a2f3d.host.com'
    })

    const app = new Application({
      name: 'A',
      version: 2,
      apdex: 55,
      contributors: [
        'John Doe'
      ],
      host: [
        'a2f3d.host.com'
      ]
    })

    Object.freeze(app)

    host.addApp(app)
  })

  it('returns an added app', () => {
    const host = new Host({
      id: 'a2f3d.host.com'
    })

    const app = new Application({
      name: 'A',
      version: 2,
      apdex: 55,
      contributors: [
        'John Doe'
      ],
      host: [
        'a2f3d.host.com'
      ]
    })

    host.addApp(app)

    const topApps = host.getTopApps(10)

    expect(topApps).toStrictEqual([app])
  })

  it('adds a smaller apdex apps in descendent order', () => {
    const host = new Host({
      id: 'a2f3d.host.com'
    })

    const appA = new Application({
      name: 'A',
      version: 2,
      apdex: 55,
      contributors: [
        'John Doe'
      ],
      host: [
        'a2f3d.host.com'
      ]
    })

    const appB = new Application({
      name: 'B',
      version: 2,
      apdex: 50,
      contributors: [
        'John Doe'
      ],
      host: [
        'a2f3d.host.com'
      ]
    })

    const appC = new Application({
      name: 'B',
      version: 2,
      apdex: 52,
      contributors: [
        'John Doe'
      ],
      host: [
        'a2f3d.host.com'
      ]
    })

    host.addApp(appA)
    host.addApp(appB)
    host.addApp(appC)

    const topApps = host.getTopApps(10)

    expect(topApps).toStrictEqual([appA, appC, appB])
  })

  it('adds a higher apdex app in descendent order', () => {
    const host = new Host({
      id: 'a2f3d.host.com'
    })

    const appA = new Application({
      name: 'A',
      version: 2,
      apdex: 70,
      contributors: [
        'John Doe'
      ],
      host: [
        'a2f3d.host.com'
      ]
    })

    const appB = new Application({
      name: 'B',
      version: 2,
      apdex: 82,
      contributors: [
        'John Doe'
      ],
      host: [
        'a2f3d.host.com'
      ]
    })

    host.addApp(appA)
    host.addApp(appB)

    const topApps = host.getTopApps(10)

    expect(topApps).toStrictEqual([appB, appA])
  })

  it('removes a single app', () => {
    const host = new Host({
      id: 'a2f3d.host.com'
    })

    const appA = new Application({
      name: 'A',
      version: 2,
      apdex: 70,
      contributors: [
        'John Doe'
      ],
      host: [
        'a2f3d.host.com'
      ]
    })

    const appB = new Application({
      name: 'B',
      version: 2,
      apdex: 30,
      contributors: [
        'John Doe'
      ],
      host: [
        'a2f3d.host.com'
      ]
    })

    const appC = new Application({
      name: 'C',
      version: 2,
      apdex: 19,
      contributors: [
        'John Doe'
      ],
      host: [
        'a2f3d.host.com'
      ]
    })

    host.addApp(appC)
    host.addApp(appA)
    host.addApp(appB)

    host.removeApp(appB)

    const topApps = host.getTopApps(10)

    expect(topApps).toStrictEqual([appA, appC])
  })

  it('removes all the apps', () => {
    const host = new Host({
      id: 'a2f3d.host.com'
    })

    const appA = new Application({
      name: 'A',
      version: 2,
      apdex: 70,
      contributors: [
        'John Doe'
      ],
      host: [
        'a2f3d.host.com'
      ]
    })

    const appB = new Application({
      name: 'B',
      version: 2,
      apdex: 30,
      contributors: [
        'John Doe'
      ],
      host: [
        'a2f3d.host.com'
      ]
    })

    const appC = new Application({
      name: 'C',
      version: 2,
      apdex: 19,
      contributors: [
        'John Doe'
      ],
      host: [
        'a2f3d.host.com'
      ]
    })

    host.addApp(appC)
    host.addApp(appA)
    host.addApp(appB)

    host.removeApp(appA)
    host.removeApp(appC)
    host.removeApp(appB)

    const topApps = host.getTopApps(10)

    expect(topApps).toStrictEqual([])
  })

  it('throws an error when attempting to remove a unexisting app', () => {
    const host = new Host({
      id: 'a2f3d.host.com'
    })

    const appA = new Application({
      name: 'A',
      version: 2,
      apdex: 19,
      contributors: [
        'John Doe'
      ],
      host: [
        'a2f3d.host.com'
      ]
    })

    expect(() => {
      host.removeApp(appA)
    }).toThrowErrorMatchingSnapshot()
  })
})
