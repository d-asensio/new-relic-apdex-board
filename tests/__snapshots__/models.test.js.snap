// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Host model throws an error when attempting to remove a unexisting app 1`] = `"The specified app do not exist in the host."`;
